### ISSUE TITLE
Lazy Async Loading Decorator differences between Node and browser

### ISSUE THREAD

**Issue by jayalfredprufrock** [mobxjs/mobx#906]
I've been trying to turn an async loading pattern I've been using into a decorator. Looks something like this: 

```
const {Component} = React;
const {observable, computed, extendObservable, useStrict, action} = mobx;
const {lazyObservable, fromPromise} = mobxUtils;
const {observer} = mobxReact;

useStrict(true);

function loadable(initialValue) {
  
  return (target, name, descriptor) => {

        const promise = descriptor.get;

        const setErrorText = action((text) => target[name + 'LoadingErrorText'] = text);

        const lazyPromise = lazyObservable((sink) => {

            const lazyPromise = fromPromise(promise().catch((error) => {

                setErrorText(error);

                return Promise.reject(initialValue);

            }), initialValue);

            sink(lazyPromise);

        }, fromPromise(Promise.resolve(initialValue)));

        const newProps = {};

        newProps[name + 'Reset'] = action(() => lazyPromise.reset());
        newProps[name + 'Refresh'] = action(() => lazyPromise.refresh());
        newProps[name + 'Loading'] = computed(() => lazyPromise.current().state == 'pending');
        newProps[name + 'LoadingError'] = computed(() => lazyPromise.current().state == 'rejected');
        newProps[name + 'LoadingErrorText'] = '';

        extendObservable(target, newProps);

        descriptor.get = () => lazyPromise.current().value;

        return descriptor;
    };
}

class Store {
  
  @loadable([])
  get something() {
    return new Promise((resolve, reject) => {
      setTimeout(() => {

        Math.random() > .25  ? resolve([1,2,3,4,5,6,7,8,9,10]) : reject('Rejection reason.');

      }, 5000);

    });
  }
   
  //if I remove the @computed decorator things work as expected
  @computed get somethingFiltered() {
     return this.something.filter((e) => e % 2);
  } 
}

@observer
class LoadableThing extends Component {

  render () {
    
    const {store} = this.props;
    
    console.log(store);
   
    if (store.somethingLoading) return <div>Loading...</div>;
    if (store.somethingLoadingError) return <div>{store.somethingLoadingErrorText}</div>;
    
    return <div>
      Original: {store.something.join(',')}<br />
      Filtered: {store.somethingFiltered.join(',')}
    </div>;
  }
}

class App extends Component {
  render = () => (
    <div>
      <LoadableThing store={ new Store() }/>
    </div>
  );
}

ReactDOM.render(<App/>, document.getElementById('container'));
```

codepen: http://codepen.io/anon/pen/XMBooq

This seems to work as expected in the codepen, and works for the most part in my node application except for one thing: the computed property that references the loadable property (`somethingFiltered` in the example) is always undefined. I've reduced my node application as much as possible to see where things are going wrong, but the behavior is always the same. This leads me to believe the issue has something to do with a difference between Node and the browser, likely involving decorators, since when I remove the loadable decorator and explicitly implement all the different pieces (and initialize the lazyObservable in the constructor), things work fine in my Node application. If I just remove the @computed decorator from the `somethingFiltered` property that also yields the correct results...

Can anybody offer a reason why this would not fully work within Node or perhaps a better pattern that I should be using to achieve the desired behavior??

**Comment by mweststrate** [comment]
Are you transpiling code / decorators with different compilers for client and server? There are subtle differences between TS and Babel for example.

**Comment by jayalfredprufrock** [comment]
I'm transpiling with Babel 6, es2015/stage0 w/ `transform-decorators-legacy`.  I'm pretty sure codepen uses a similar setup, but obviously it's pulling in all the dependencies from CDNs.

**Comment by andykog** [comment]
@jayalfredprufrock, I've tried using your example in node environment (SSR), seems to work fine, check this branch: https://github.com/andykog/mobx-react-todomvc/tree/906

**Comment by jayalfredprufrock** [comment]
Thanks for looking into this @andykog. In the end I opted for more predictable control over loading than the lazy approach I was originally going for since it felt like I was bordering on an anti-pattern. My updated version is cleaner, more predictable, _and_ doesn't suffer from whatever issue I was hitting before. I'll go ahead and close this, but I updated my original code (and the codepen) in case others find it useful. Thanks again to all!
